File: Cpp/1496_Path_Crossing.cpp
class Solution {
public:
    bool isPathCrossing(string path) {
        int x = 0;
        int y = 0;
        unordered_map<int, unordered_set<int>> visit;
        visit[0].insert(0);
        
        for(auto& direction : path){
            if(direction == 'N'){
                --y;
            } else if(direction == 'S'){
                ++y;
            } else if(direction == 'E'){
                ++x;
            } else if(direction == 'W'){
                --x;
            }
            if(visit[x].find(y) != visit[x].end()) return true;
            visit[x].insert(y);
        }
        
        return false;
    }
};

File: Kotlin/1496_Path_Crossing.kt
class Solution {
    fun isPathCrossing(path: String): Boolean {
        var x = 0
        var y = 0
        val visit = mutableSetOf("$x,$y")
        
        for(direction in path){
            when (direction) {
                'N' -> --y
                'S' -> ++y
                'E' -> ++x
                'W' -> --x
            }
            if (visit.contains("$x,$y")) return true
            visit.add("$x,$y")
        }
        
        return false
    }
}